thewalrus/loop_hafnian_batch.py

```diff
@@ -0,0 +1,297 @@
+# Copyright 2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Sampling algorithms
+===================
+
+**Module name:** :mod:`thewalrus.loop_hafnian_batch`
+
+.. currentmodule:: thewalrus.loop_hafnian_batch
+
+Tools to evaluate the loop hafnian batch to do gaussian boson sampling.
+
+Summary
+-------
+
+.. autosummary::
+    _calc_loop_hafnian_batch_even
+    _calc_loop_hafnian_batch_odd
+    add_batch_edges_even
+    add_batch_edges_odd
+    loop_hafnian_batch
+
+Code details
+------------
+"""
+import numpy as np
+import numba
+from thewalrus._hafnian import (
+    precompute_binoms,
+    matched_reps,
+    find_kept_edges,
+    f_loop,
+    f_loop_odd,
+    get_submatrices,
+    get_submatrix_batch_odd0,
+)
+
+# pylint: disable = too-many-arguments, not-an-iterable
+@numba.jit(nopython=True, parallel=True, cache=True)
+def _calc_loop_hafnian_batch_even(A, D, fixed_edge_reps, batch_max, odd_cutoff, glynn=True):
```

Comment:
```suggestion
def _calc_loop_hafnian_batch_even(A, D, fixed_edge_reps, batch_max, odd_cutoff, glynn=True):  # pragma: no cover
```

Reply:
Oh it makes sense! We will try it out. Thanks!